core/src/main/java/io/grpc/internal/AbstractManagedChannelImplBuilder.java

```diff
@@ -241,11 +241,18 @@ public ManagedChannelImpl build() {
       transportFactory = new AuthorityOverridingTransportFactory(
         transportFactory, authorityOverride);
     }
+    NameResolver.Factory nameResolverFactory = this.nameResolverFactory;
+    if (nameResolverFactory == null) {
+      // Avoid loading the provider unless necessary, as a way to workaround a possibly-costly
+      // and poorly optimized getResource() call on Android. If any other piece of code calls
+      // getResource(), then this shouldn't be a problem unless called on the UI thread.
+      nameResolverFactory = NameResolverProvider.asFactory();
+    }
```

Comment:
MCB isn't referenced from any library code. Only the application calls it. So they simply would use the okhttp or cronet builder directly (which is what I assume they are doing already, otherwise they wouldn't have noticed this).

We will want to have a fix for MCB and friends though, but we can work on that separately from stopping the bleeding.